GWAS data handling improvements

kammerer opened the pull request:
Fixes #721 
Fixes #722 
Fixes #726 

Changes:

- make sure line breaks in input files are consistent and CSV parser works correctly
- make sure Manhattan plot handles non-numeric chromosome ids/names correctly
- replace special characters in mutation, sample and trait names with underscores because GWASSER does not handle them well
- remove SNPs which do not vary between samples (only one distinct value except for NA) from the input
- remove traits which do not vary between samples from the input (uploaded or based on selected PT)
- show removed SNPs and traits to the user
- load GWAS results asynchronously
- do not plot SNPs with -log10(p) below specified cutoff (for data sets > 10000 samples)

Nuanda commented:
@kammerer I can do the review tomorrow or the day after tomorrow the earliest. Is that ok with you?

kammerer commented:
Sure, no problem.

2017-06-13 14:02 GMT+02:00 Tomasz Gubała <notifications@github.com>:

> @kammerer <https://github.com/kammerer> I can do the review tomorrow or
> the day after tomorrow the earliest. Is that ok with you?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/TGAC/brassica/pull/725#issuecomment-308094275>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAA2ySf80j4Q0rMfJtDPYBsIDY8cMEm9ks5sDnpigaJpZM4N0Z9h>
> .
>



-- 
Tomasz Szymczyszyn

Nuanda commented on app/services/analysis/gwas/genotype_vcf_to_csv_converter.rb:
```diff
@@ -52,6 +35,79 @@ def call(path)
+      #   0 for no mutation
+      #   NA if call could not be made
+      def sample_value(sample, alternative_no)
+        if sample.empty?
```
Maybe `return "NA" if sample.empty?` (also remove the comment, since it is already explained above the signature).

Nuanda commented on app/services/analysis/gwas/setup.rb:
```diff
@@ -0,0 +1,129 @@
+          end
+        end
+
+        values_by_mutation.map { |mutation, values| mutation if (values - ["NA"]).size < 2 }.compact - ["ID"]
```
Split into multiple lines?

kammerer commented on app/controllers/analyses_controller.rb:
```diff
@@ -32,12 +32,26 @@ def show
+      format.json do
+        if @analysis.gwas? && @analysis.success?
+          results = Analysis::Gwas::ManhattanPlot.new(@analysis).call.fetch(:traits).map do |trait_name, mutations, _|
+            mutations.map { |m| m[1] = "%.4f" % m[1]; m << trait_name }
```
I am appending trait name to each mutation in the list (this is content for the results table).

kammerer commented:
@Nuanda These are files uploaded by Annemarie, 60k+ SNPs are included. Analysis takes several hours on my box:

[Pheno file](https://drive.google.com/open?id=0BzlTet3SNRNTNDlyeUNDcEVkZWM)
[Geno file](https://drive.google.com/open?id=0BzlTet3SNRNTR2pHeXU1U0pmRU0)
[Map file](https://drive.google.com/open?id=0BzlTet3SNRNTRGN3VFE0U1dPUWM)

As for PT, I used RIPR root, generated templates for geno and map and altered them a bit. I am not sharing these files as my local PT data is incomplete and would probably not match yours.

Nuanda commented:
Is the generated genotype data template supposed to have the PSUs from the selected Plant Trial? It doesn't in my case.

Nuanda commented on app/services/analysis/gwas/manhattan_plot.rb:
```diff
@@ -1,58 +1,70 @@
-            tooltips = mutations.map { |mut| format_tooltip(*mut) }
+            mutations.each_with_index do |(_, _, chrom, _), idx|
+              chromosomes[chrom] << idx
+            end if mutations.first.size > 2
```
I tried to run GWAS with a very simple input (2 PSUs, 1 trait, generated geno and map files) and I get

```undefined method `size' for nil:NilClass```

 in this line when trying to see the results. The analysis finished with success.

kammerer commented:
Yes, the template should have PSUs from the selected Plant Trial. Does the template download link have `plant_trial_id` parameter?

Nuanda commented:
You are correct. Apparently I haven't reloaded my server after switching branches, and somehow an old generator class was still loaded. I retried that and it generates expected PSU names.

Nuanda commented:
Trying with another, also small, PT, I had the analysis finish with error. The STD error reads:

```
Now reading genotype file...
Error in `[.data.frame`(geno, , i) : undefined columns selected
Calls: parseGeno -> as.factor -> is.factor -> [ -> [.data.frame
Execution halted
```

Do you know what might be the cause? In case you don't, I'll try to supply you with my input files.

kammerer commented:
I don't think I have seen such an error from GWASSER. Let me have a look at the files.

Nuanda commented:
Here they come. Pheno was created from a PT I submitted previously:

[genotype.csv.txt](https://github.com/TGAC/brassica/files/1080477/genotype.csv.txt)
[map.csv.txt](https://github.com/TGAC/brassica/files/1080476/map.csv.txt)
[pheno.csv.txt](https://github.com/TGAC/brassica/files/1080478/pheno.csv.txt)

Nuanda commented:
Another GWAS failure case, this time slightly more complex. I used a CS-old PT, `whri_2005_GE2_02`, and uploaded generated geno and map files, with no changes. The following failure happened:

```
--------------------------------------------
 gwasseR: an R script for simple gwassing  
--------------------------------------------
Adapted from the work of researchers at NIAB
Techniques here were used for MAGIC analysis of awning

Input genotype file:
  /home/gubala/ini/bip/brassica/public/system/analysis/data_files/30e5cfccf8d7dd9f1bd806ac06ae74a045199b4e.csv

Input phenotype files:
  /home/gubala/ini/bip/brassica/public/system/analysis/data_files/f2e8196272af622d7d44f7d26c2e48505071b5d2.csv

Directory for output files:
 /home/gubala/ini/bip/brassica/tmp/exec/development/12/results

Using Phenotypes:
  shoot_percent_dry_weight
  sodium_content
  shoot_magnesium_content_(Mg)
  zinc_content
  shoot_boron_content_(B)
  shoot_calcium_content_(Ca)
  shoot_carbon_content_(C)_
  copper_content
  shoot_dry_weight
  shoot_fresh_weight
  iron_content
  manganese_content
  nitrogen_content
  phosphorus_content
  potassium_content

Using Map:
 

Using mixed effects:
  

------------------------
Start Analysis:

Now reading genotype file...
Now reading phenotype files...
Merging genotype and phenotype files...
Starting association of SNPs for phenotype: shoot_percent_dry_weight
Error in parse(text = x, keep.source = FALSE) : 
  <text>:7:25: unexpected input
6: shoot_calcium_content_(Ca) ~  + ID.1
7: shoot_carbon_content_(C)_
                           ^
Calls: snpAssociation ... formula -> formula.character -> formula -> eval -> parse
Execution halted
```

Input files:
[genotype2.csv.txt](https://github.com/TGAC/brassica/files/1080483/genotype2.csv.txt)
[map2.csv.txt](https://github.com/TGAC/brassica/files/1080484/map2.csv.txt)
[pheno2.csv.txt](https://github.com/TGAC/brassica/files/1080485/pheno2.csv.txt)

kammerer commented:
Thanks for the files. The first trial fails because there is too little data to fit the model. All mutations were removed from input because there was no difference between samples (NAs do not count). One of the traits has the same issue. I need to think how to handle that yet.
The second one looks like an error in trait name normalization.

I will let you know when I fix these.

kammerer commented:
@Nuanda I fixed the issues you found and added more specs for such cases. I intentionally left some TODOs as I will continue to work on this but I think current changes are worth merging to the master. Besides this PR is already too big.

Nuanda commented on app/services/analysis/gwas/manhattan_plot.rb:
```diff
@@ -1,28 +1,30 @@
       def call
-        { traits: [], chromosomes: [], cutoff: cutoff }.tap do |results|
-
+        { traits: [], chromosomes: [] }.tap do |results|
```
You can drop :chromosomes key init as well.

Nuanda commented:
I assume Genotype data and genotype map upload buttons are drop-enabled. I can drop a file there, however, when I drop a genotype file, it updates both upload fields at once (the genotype map file is updated as well).

Nuanda commented:
Both testing cases behave correctly now. Just two considerations - one syntax, one UX. Good to merge.

kammerer commented:
@Nuanda Thanks, fixed both issues. Merging.
